### ISSUE TITLE
Add ability to harden standard accounts

### ISSUE THREAD

**Issue by Dyras** [securitywithoutborders/hardentools#31]
Greetings! First of all, great program. However, it was come to my attention that if I run the program from a standard account it asks me for admin privileges. Great, so I enter my admin password and I run the tool. All seems well.

However, on my standard account I can still open CMD. This to me makes it seem as if hardening standard accounts is impossible with this program. Considering Microsoft recommends using a standard account for day-to-day business, that seems odd.

So tl;dr I wish I could harden standard accounts as well. I'm mostly interested in the Office hardening though.

**Comment by obsti8383** [comment]
Hi Dyras, you are right, hardening standard accounts is not supported right now (that should be documented in the readme). Background here is, that the language GO doesn't support to request elevated rights in the application itself. So we would have to deliver to different executables to accomplish that. Since we assumed our primary target group uses local admin rights with their primary account, we kept it that way for now. 
I would propose to keep that ticket open, since hardening standard accounts is also a thing I would like to have. But we need a better logic (for GUI and detection of current setup/settings) first, before we can publish two different version without too much overhead.

**Comment by Velocet** [comment]
Maybe it would be better to re-write the app as a PowerShell script? This way you could invoke C# when there is no direct PoSh function and you could also build a .exe for distribution (like with [PS2EXE](https://gallery.technet.microsoft.com/PS2EXE-Convert-PowerShell-9e4e07f1)).

This way it would also be more easy for admins to locally change the script and everything would be native. Thanks to Visual Studio Community (and other tools) it is also possible to build a GUI for this script.

**Comment by Dyras** [comment]
@obsti8383 Disabling Windows Script Host can easily be done for all users at least:
https://technet.microsoft.com/en-us/library/ee198684.aspx

**Comment by botherder** [comment]
The irony of rewriting it in PowerShell is evil, but it is true that in retrospect, Golang isn't perhaps the best choice for this. The reason why I picked go in the first place was because I had envisioned to make this cross-platform and implement appropriate changes for other operating systems too in a single codebase. Given that it is all boiling down to Windows registry key changes, I'm open to revisit this choice.

**Comment by botherder** [comment]
I'll move this conversation to a different ticket.

**Comment by dragokas** [comment]
@obsti8383 
> Background here is, that the language GO doesn't support to request elevated rights in the application itself.

What you mean? (ShellExecute + 'runas'?) Is it just to identify what user initially run the hardening?
You can use ProcessIdToSessionId + WTSQuerySessionInformation

Also, why you just can't add checkbox "v Apply to ALL users" and process every currently available HKU\<SID> hive?

**Comment by obsti8383** [comment]
@dragokas : I currently tend to implement your second solution (since this fixes settings for all accounts). However since it would also be nice to allow the parts that need no admin rights to be executed by unprivileged users, do you have a code sample in Golang that shows how to call ShellExecute and the user identification part?

**Comment by dragokas** [comment]
You don't need ShellExecute.
If you want to run your app under account that belongs to 'Users' group without elevation, all you need:

replace:
<requestedExecutionLevel level="requireAdministrator" uiAccess="false"/>
by:
<requestedExecutionLevel level="highestAvailable" uiAccess="false"/>
in your manifest file.

P.S. I don't have any experience in GoLang.

**Comment by obsti8383** [comment]
Took the first step successfully for this feature. My current release in my repository (https://github.com/obsti8383/hardentools/releases/tag/0.4) now doesn't request admin privileges in the manifest file, but pops up a dialog if it is run with normal user privileges:
![yesno](https://user-images.githubusercontent.com/16605975/43370113-812fffbc-9379-11e8-8aa4-b751e20c5b3f.png)

If the user selects "No", he only gets the subset of hardening methods that are available without admin rights (Office and so on):
![standarduser](https://user-images.githubusercontent.com/16605975/43370122-b321567e-9379-11e8-9bd0-a43dd0ed307d.png)

However that still does not address hardening all users via a "Apply to all users" feature.

If you want to compile from source, please be aware, that you need gcc from MingW (for the windows includes) installed and available in path (package i686-w64-mingw32-gcc on Linux) due to inline C code in main.go for starting a new hardentools.exe instance via ShellExecuteEx().

Feedback welcome.

**Comment by botherder** [comment]
That looks great!

**Comment by obsti8383** [comment]
Fixed by #66

**Comment by obsti8383** [comment]
@botherder : You can close this one (fixed with #66)
